src/clojure/nrepl/rich_printer.clj

```diff
@@ -0,0 +1,460 @@
+(ns nrepl.rich-printer
+  "A printer that produces a pure EDN representation of a Clojure value.
+  Features elisions, MIME attachments and general purpose object browsing.
+  Initially developed for unrepl."
+  (:refer-clojure :exclude [pr pr-str])
+  (:require
+   [clojure.core :as core]
+   [clojure.string :as str]
+   [clojure.edn :as edn]
+   [clojure.main :as main]))
+
+; even with limits on breadth and depth of printing, we can easily get big values:
+; the biggest tree would have fitting in these lilites would have breadth-limit^depth-limit
+; so even with small defaults such as 8 for depth and 10 for breadth we get 100 millions of leaves.
+; That's why we introduce the notion of budget, which is breadth-limits*depth-limit.
+; (this is a heuristic which could be tuned)
+; Once the budget is exhausted everything is elided.
+(def ^:dynamic ^:private *print-budget*)
+
+(def ^:dynamic *elide*
+  "Function of one argument (the value to elide) which returns its elision
+  (as a map with at least a :get key containing a form to evaluate to retrieve the elided value)
+  or nil when the elision is definitive."
+  (constantly nil))
+
+(def ^:dynamic *realize-on-print*
+  "Set to false to avoid realizing lazy sequences."
+  true)
+
+(def ^:dynamic *string-length* 72)
+
+(def defaults {#'*print-length* 10
+               #'*print-level* 8
+               #'*string-length* 72})
+
+(defn- bump [n m]
+  (if (< n (- Long/MAX_VALUE m))
+    (+ n m)
+    Long/MAX_VALUE))
+
+(defn ensure-defaults [bindings]
+  (let [bindings (merge-with #(or %1 %2) bindings defaults)]
+    (assoc bindings #'*print-budget*
+           (long (min (* 1N (bindings #'*print-level*) (bindings #'*print-length*)) Long/MAX_VALUE)))))
+
+(defprotocol MachinePrintable
+  (-print-on [x write rem-depth]))
+
+;; clojure 1.10 support
```

Comment:
Or datafy/nav support? :-) Probably this will be a better description once 1.11 is out. :-)